File: examples/companion-packages/haskell-hackage-app/src/AgentXM/Example/PawMatch/Flags.hs
{-# LANGUAGE OverloadedStrings #-}

-- |
-- Module      : AgentXM.Example.PawMatch.Flags
-- Description : Flag-name constants and registry for the PawMatch app.
module AgentXM.Example.PawMatch.Flags
  ( homeCheckFollowup
  , feeBreakdownDetailed
  , longStayHighlight
  , suggestDonateAfterAdoption
  , showCharityRatings
  , recommendationStrategy
  , matchQuizDepth
  , petCardStyle
  , donateFocusDefault
  , buildRegistry
  ) where

import           AgentXM.Example.TinyFlags
import           Data.Text (Text)

homeCheckFollowup, feeBreakdownDetailed, longStayHighlight :: Text
homeCheckFollowup           = "home-check-followup"
feeBreakdownDetailed        = "fee-breakdown-detailed"
longStayHighlight           = "long-stay-highlight"

suggestDonateAfterAdoption, showCharityRatings :: Text
suggestDonateAfterAdoption  = "suggest-donate-after-adoption"
showCharityRatings          = "show-charity-ratings"

recommendationStrategy, matchQuizDepth, petCardStyle, donateFocusDefault :: Text
recommendationStrategy      = "recommendation-strategy"
matchQuizDepth              = "match-quiz-depth"
petCardStyle                = "pet-card-style"
donateFocusDefault          = "donate-focus-default"

-- | Construct the PawMatch flag registry. Smart-constructor errors crash
-- early via 'error' because these definitions are static — a misconfigured
-- registry indicates a programmer error, not a recoverable condition.
buildRegistry :: Registry
buildRegistry =
  registry
    [ (homeCheckFollowup,          FBool    (mustBool False (Just 25)))
    , (feeBreakdownDetailed,       FBool    (mustBool True  Nothing))
    , (longStayHighlight,          FBool    (mustBool True  Nothing))
    , (suggestDonateAfterAdoption, FBool    (mustBool False (Just 50)))
    , (showCharityRatings,         FBool    (mustBool True  Nothing))
    , ( recommendationStrategy
      , FVariant
          (mustVariant
            ["popularity", "match-quiz", "longest-stay"]
            "match-quiz"
            (Just [("longest-stay", 20)]))
      )
    , ( matchQuizDepth
      , FVariant
          (mustVariant
            ["short", "standard", "thorough"]
            "standard"
            Nothing)
      )
    , ( petCardStyle
      , FVariant
          (mustVariant
            ["compact", "detailed", "playful"]
            "detailed"
            Nothing)
      )
    , ( donateFocusDefault
      , FVariant
          (mustVariant ["all", "shelters", "rescue"] "all" Nothing)
      )
    ]
  where
    mustBool d r =
      case booleanFlag d r of
        Right f  -> f
        Left err -> error ("PawMatch buildRegistry: invalid boolean flag — " <> show err)
    mustVariant vs d r =
      case variantFlag vs d r of
        Right f  -> f
        Left err -> error ("PawMatch buildRegistry: invalid variant flag — " <> show err)

File: examples/companion-packages/maven-java-app/src/main/java/ai/agentxm/examples/pawmatch/Flags.java
package ai.agentxm.examples.pawmatch;

import ai.agentxm.examples.tinyflags.TinyFlags;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/** Flag keys and the canonical PawMatch {@link TinyFlags} bundle. */
public final class Flags {

    public static final String HOME_CHECK_FOLLOWUP = "home-check-followup";
    public static final String FEE_BREAKDOWN_DETAILED = "fee-breakdown-detailed";
    public static final String LONG_STAY_HIGHLIGHT = "long-stay-highlight";
    public static final String SUGGEST_DONATE_AFTER_ADOPTION = "suggest-donate-after-adoption";
    public static final String SHOW_CHARITY_RATINGS = "show-charity-ratings";
    public static final String RECOMMENDATION_STRATEGY = "recommendation-strategy";
    public static final String MATCH_QUIZ_DEPTH = "match-quiz-depth";
    public static final String PET_CARD_STYLE = "pet-card-style";
    public static final String DONATE_FOCUS_DEFAULT = "donate-focus-default";

    private Flags() {}

    public static TinyFlags create() {
        Map<String, Integer> recommendationRollout = new LinkedHashMap<>();
        recommendationRollout.put("longest-stay", 20);

        return TinyFlags.builder()
                .booleanFlag(HOME_CHECK_FOLLOWUP, false, 25)
                .booleanFlag(FEE_BREAKDOWN_DETAILED, true)
                .booleanFlag(LONG_STAY_HIGHLIGHT, true)
                .booleanFlag(SUGGEST_DONATE_AFTER_ADOPTION, false, 50)
                .booleanFlag(SHOW_CHARITY_RATINGS, true)
                .variantFlag(
                        RECOMMENDATION_STRATEGY,
                        List.of("popularity", "match-quiz", "longest-stay"),
                        "match-quiz",
                        recommendationRollout)
                .variantFlag(
                        MATCH_QUIZ_DEPTH,
                        List.of("short", "standard", "thorough"),
                        "standard")
                .variantFlag(
                        PET_CARD_STYLE,
                        List.of("compact", "detailed", "playful"),
                        "detailed")
                .variantFlag(
                        DONATE_FOCUS_DEFAULT,
                        List.of("all", "shelters", "rescue"),
                        "all")
                .build();
    }
}
